src/io/mod.rs

```diff
@@ -26,8 +26,101 @@ pub struct Obfuscation<M: PolyMatrix> {
     #[cfg(feature = "test")]
     pub final_preimage_target: M,
 }
+#[cfg(test)]
+mod test {
+    use std::sync::Arc;
+
+    use keccak_asm::Keccak256;
+    use num_bigint::BigUint;
+    use rand::Rng;
+    use tracing::info;
+
+    use crate::{
+        bgg::circuit::PolyCircuit,
+        io::{obf::obfuscate, params::ObfuscationParams},
+        poly::{
+            dcrt::{
+                DCRTPoly, DCRTPolyHashSampler, DCRTPolyMatrix, DCRTPolyParams,
+                DCRTPolyTrapdoorSampler, DCRTPolyUniformSampler, FinRingElem,
+            },
+            sampler::DistType,
+            Poly, PolyElem, PolyParams,
+        },
+    };
 
-// impl<M: PolyMatrix> Obfuscation<M> {
-//     pub fn read_from_files
-// }
-// [TODO] Add a function to read/write the obfuscation to/from files
+    const SIGMA: f64 = 4.578;
+
+    #[test]
+    fn test_io_just_mul_enc_and_and_bit() {
+        let start_time = std::time::Instant::now();
+        let params = DCRTPolyParams::new(4, 2, 17, 10);
+        let log_q = params.modulus_bits();
+        let switched_modulus = Arc::new(BigUint::from(1u32));
+        let mut public_circuit = PolyCircuit::new();
+
+        // inputs: BITS(ct), eval_input
+        // outputs: BITS(ct) AND eval_input, BITS(ct) AND eval_input
+        {
+            let inputs = public_circuit.input((2 * log_q) + 1);
+            let mut outputs = vec![];
+            let eval_input = inputs[2 * log_q];
+            for ct_input in inputs[0..2 * log_q].iter() {
+                let muled = public_circuit.and_gate(*ct_input, eval_input);
+                outputs.push(muled);
+            }
+            for ct_input in inputs[0..2 * log_q].iter() {
+                let muled = public_circuit.and_gate(*ct_input, eval_input);
+                outputs.push(muled);
+            }
+            public_circuit.output(outputs);
+        }
+
+        let obf_params = ObfuscationParams {
+            params: params.clone(),
+            switched_modulus,
+            input_size: 1,
+            public_circuit: public_circuit.clone(),
+            d: 3,
+            encoding_sigma: 0.0,
+            hardcoded_key_sigma: 3.0,
```

Comment:
Could you set `hardcoded_key_sigma ` to 0.0 for the dummy params to check that intermediate data are computed correctly during evaluation?